lib/ReplaceOpenCLBuiltinPass.cpp

```diff
@@ -2683,6 +2687,105 @@ bool ReplaceOpenCLBuiltinPass::replaceVstoreHalf4(Module &M) {
   });
 }
 
+bool ReplaceOpenCLBuiltinPass::replaceUnsampledReadImage(Module &M) {
+  bool Changed = false;
+  const std::map<const char *, const char *> Map = {
+      // 1D
+      {"_Z11read_imagef14ocl_image1d_roi",
+       "_Z11read_imagef14ocl_image1d_ro11ocl_sampleri"},
+      {"_Z11read_imagei14ocl_image1d_roi",
+       "_Z11read_imagei14ocl_image1d_ro11ocl_sampleri"},
+      {"_Z12read_imageui14ocl_image1d_roi",
+       "_Z12read_imageui14ocl_image1d_ro11ocl_sampleri"},
+      // TODO 1D array
+      // 2D
+      {"_Z11read_imagef14ocl_image2d_roDv2_i",
+       "_Z11read_imagef14ocl_image2d_ro11ocl_samplerDv2_i"},
+      {"_Z11read_imagei14ocl_image2d_roDv2_i",
+       "_Z11read_imagei14ocl_image2d_ro11ocl_samplerDv2_i"},
+      {"_Z12read_imageui14ocl_image2d_roDv2_i",
+       "_Z12read_imageui14ocl_image2d_ro11ocl_samplerDv2_i"},
+      // TODO 2D array
+      // 3D
+      {"_Z11read_imagef14ocl_image3d_roDv4_i",
+       "_Z11read_imagef14ocl_image3d_ro11ocl_samplerDv4_i"},
+      {"_Z11read_imagei14ocl_image3d_roDv4_i",
+       "_Z11read_imagei14ocl_image3d_ro11ocl_samplerDv4_i"},
+      {"_Z12read_imageui14ocl_image3d_roDv4_i",
+       "_Z12read_imageui14ocl_image3d_ro11ocl_samplerDv4_i"}};
+
+  Function *translate_sampler =
+      M.getFunction(clspv::TranslateSamplerInitializerFunction());
+  Type *sampler_type = M.getTypeByName("opencl.sampler_t");
+  for (auto Pair : Map) {
+    // If we find a function with the matching name.
+    if (auto F = M.getFunction(Pair.first)) {
+      SmallVector<Instruction *, 4> ToRemoves;
+
+      // Walk the users of the function.
+      for (auto &U : F->uses()) {
+        if (auto CI = dyn_cast<CallInst>(U.getUser())) {
+          // The image.
+          auto Image = CI->getOperand(0);
+
+          // The coordinate.
+          auto Coord = CI->getOperand(1);
+
+          // Create the sampler translation function if necessary.
+          if (!translate_sampler) {
+            // Create the sampler type if necessary.
+            if (!sampler_type) {
+              sampler_type =
+                  StructType::create(M.getContext(), "opencl.sampler_t");
+              sampler_type = sampler_type->getPointerTo(2);
```

Comment:
meh. 2 is constant address space, right...

Reply:
Good point. Fixed.